source/material_model/visco_plastic.cc

```diff
@@ -280,6 +280,23 @@ namespace aspect
           // Calculate changes in strain invariants and update the reaction terms
           rheology->strain_rheology.fill_reaction_outputs(in, i, rheology->min_strain_rate, plastic_yielding, out);
 
+          // Calculate changes in viscosity with iterative dampening and update the reaction terms
+          if (rheology->use_iterative_viscosity_dampening)
+            {
+              if (in.current_cell.state() == IteratorState::valid)
+                {
+                  const double old_viscosity = in.composition[i][this->introspection().compositional_index_for_name("viscosity_field")];
+
+                  // Only dampen after the first nonlinear iteration
+                  if (this->get_nonlinear_iteration() > 0)
+                    out.viscosities[i] = rheology->iterative_dampening->calculate_viscosity(old_viscosity, out.viscosities[i]);
+
+                  rheology->iterative_dampening->fill_reaction_outputs(in, i, old_viscosity, out);
```

Comment:
I had a brief discussion with @anne-glerum and @bobmyhill about this, so just to document this here: I would use the PrescribedFieldOutputs here instead of reaction outputs. Make your viscosity field a prescribed field instead of solving an equation for it in every time step (that is cheaper anyway). You can then compute your update from the viscosity in the previous iteration stored in `in.composition[i][this->introspection().compositional_index_for_name("viscosity_field")]` and the new viscosity in `out.viscosities[i]`. 

Instead of solving, you would then just copy your prescribed field outputs into your compositional field (no reaction terms required). That happens in the `interpolate_material_output_into_advection_field` function. I just checked again and `in.composition` in this case is based on the values in the solution vector, and the solution vector for your prescribed field would be updated in this function as well (directly after you have computed your update). So that would solve your issues with the updates. 

For an example, see this file:
https://github.com/geodynamics/aspect/blob/735ef46d8ece46075d5413f467c21c577817c3c9/tests/multiple_named_additional_outputs.cc#L51-L67

Reply:
@jdannberg - fantastic, thank you! The approach you outlined should do the trick.